### PR TITLE
Generate additional Windows Server templates specific to Windows 2k12 R2, Windows 2k16 and Windows 2k19

### PR DESCRIPTION
**Pull request by shwetaap**
**What this PR does / why we need it**:
This PR generates additional Windows Server templates specific to Windows 2k12 R2, Windows 2k16 and Windows 2k19
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
None
```

### CODE REVIEW

File: templates/windows2k12.tpl.yaml [hunk 1]
```diff
@@ -2,11 +2,11 @@
 apiVersion: template.openshift.io/v1
```
Comment (omeryahud) [inline templates/windows2k12.tpl.yaml:2]: All new templates you created should have `{% set version = "0.12.3" %}` as this is going to be the first version they were introduced into common-templates

**Review by omeryahud — CHANGES_REQUESTED** [review]

File: templates/windows2k12.tpl.yaml [hunk 2]
```diff
@@ -149,7 +145,7 @@ parameters:
   required: true
 - name: SRC_PVC_NAME
   description: Name of the PVC to clone
-  value: win2k19
+  value: win2k12r2
```
[the remark this answers is not in the document]
Comment (omeryahud) [reply templates/windows2k12.tpl.yaml:148]: Nevermind, `win2kr12` is better

File: automation/test-windows.sh [hunk 3]
```diff
@@ -90,7 +90,7 @@ sizes=("medium" "large")
 workloads=("server")
 
 if [[ $TARGET =~ windows10.* ]]; then
```
[the remark this answers is not in the document]
Comment (omeryahud) [reply automation/test-windows.sh:92]: You can see the exact `$TARGET` values from the command line specified in the e2e prow job here https://github.com/kubevirt/project-infra/blob/master/github/ci/prow/files/jobs/kubevirt/common-templates/common-templates-presubmits.yaml

File: automation/test-windows.sh [hunk 4]
```diff
@@ -194,9 +194,11 @@ run_vm(){
   fi
 }
 
-for size in ${sizes[@]}; do
-  for workload in ${workloads[@]}; do
-    vm_name=$template_name-$workload-$size
-    run_vm $vm_name
+for template_name in ${template_names[@]}; do
```
[the remark this answers is not in the document]
Comment (shwetaap) [reply automation/test-windows.sh]: Done

File: templates/windows2k12.tpl.yaml (hunk 1 above)
Comment (shwetaap) [reply templates/windows2k12.tpl.yaml:2]: Done

File: automation/test-windows.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (shwetaap) [reply automation/test-windows.sh:92]: Done

File: automation/test-windows.sh [hunk 5]
```diff
@@ -6,7 +6,7 @@ _oc() {
   cluster-up/kubectl.sh "$@"
 }
 
-template_name="windows"
+template_names=("windows2k12r2" "windows2k16" "windows2k19")
```
[the remark this answers is not in the document]
Comment (shwetaap) [reply automation/test-windows.sh]: Done

File: travis_ci/test_duplicate_templates.sh [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Copyright 2019 Red Hat, Inc.
+#
+set -ex
+
+make generate
```
Comment (omeryahud) [inline travis_ci/test_duplicate_templates.sh:21]: Travis fails this test because you only generated the templates locally and haven't applied them to the cluster.

File: templates/windows2k12.tpl.yaml [hunk 7]
```diff
@@ -63,9 +61,7 @@ metadata:
       ]
 
   labels:
-    os.template.kubevirt.io/win2k19: "true"
-    os.template.kubevirt.io/win2k16: "true"
-    os.template.kubevirt.io/win2k12r2: "true"
+    os.template.kubevirt.io/windows2k12r2: "true"
```
[the remark this answers is not in the document]
Comment (omeryahud) [reply templates/windows2k12.tpl.yaml:64]: This should be `os.template.kubevirt.io/win2k12r2: "true"`

File: templates/windows10.tpl.yaml [hunk 8]
```diff
@@ -61,7 +61,7 @@ metadata:
       ]
 
   labels:
-    os.template.kubevirt.io/win10: "true"
+    os.template.kubevirt.io/windows10: "true"
```
[the remark this answers is not in the document]
Comment (omeryahud) [reply templates/windows10.tpl.yaml:64]: Please revert it to `os.template.kubevirt.io/win10: "true"`

File: .travis.yml [hunk 9]
```diff
@@ -39,6 +39,9 @@ jobs:
   - name: Functional test of the CentOS 6 template
     script: bash -x travis_ci/test.sh centos6 server
 
+  - name: Test to check if there are duplicate templates create
```
Comment (omeryahud) [inline .travis.yml:42]: I would reword this to "Ensure that a single template is exposed for each template variant (os/workload/size trio)"

**Review by omeryahud — CHANGES_REQUESTED** [review]

File: templates/windows10.tpl.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (shwetaap) [reply templates/windows10.tpl.yaml:64]: The reason I changed this was that the I wanted the label to include a prefix of the OS(eg "os.template.kubevirt.io/centos6") and not just "os.template.kubevirt.io/". This prefix is the same as the OS prefix in the filename of the template(eg: centos6-server-large.yaml) for all templates except the windows templates, where the label has "win*", but the filename prefix is "windows*". So I tried changing the label to match the os prefix in the filename of the template. 
I could do either of the two options if the label change is not allowed 
1. rename the template filenames to match the label or
2. Add a different "if" clause for the windows templates or
3. Do not use the OS name in while matching the label

File: .travis.yml (hunk 9 above)
Comment (shwetaap) [reply .travis.yml:42]: Done

File: templates/windows10.tpl.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (shwetaap) [reply templates/windows10.tpl.yaml:64]: Done

File: templates/windows2k12.tpl.yaml (hunk 7 above)
[the remark this answers is not in the document]
Comment (shwetaap) [reply templates/windows2k12.tpl.yaml:64]: Done

File: templates/windows2k16.tpl.yaml [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+          "path": "jsonpath::.spec.domain.devices.disks[*].cdrom.bus",
+          "valid": "jsonpath::.spec.domain.devices.disks[*].cdrom.bus",
+          "rule": "enum",
+          "message": "cd bus has to be sata",
+          "values": ["sata"]
+        }
+      ]
+
+  labels:
+    os.template.kubevirt.io/windows2k16: "true"
```
[the remark this answers is not in the document]
Comment (shwetaap) [reply templates/windows2k16.tpl.yaml:64]: Done

File: templates/windows2k19.tpl.yaml [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+          "path": "jsonpath::.spec.domain.devices.disks[*].cdrom.bus",
+          "valid": "jsonpath::.spec.domain.devices.disks[*].cdrom.bus",
+          "rule": "enum",
+          "message": "cd bus has to be sata",
+          "values": ["sata"]
+        }
+      ]
+
+  labels:
+    os.template.kubevirt.io/windows2k19: "true"
```
[the remark this answers is not in the document]
Comment (shwetaap) [reply templates/windows2k19.tpl.yaml:64]: Done

File: travis_ci/test_duplicate_templates.sh [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+
+templates=("dist/templates/*.yaml")
+for template in $templates; do
+   templatename=$(echo $template | cut -d'/' -f3 | cut -d'-' -f1)
+
+   # Different labels for workload, flavor, and template is not found in the deprecated template - win2k12r2-deprecated.tpl.yaml
+   if [[ $templatename != "win2k12r2" ]]; then
+       template_name_label="os.template.kubevirt.io/${templatename}.*true.*"
```
[the remark this answers is not in the document]
Comment (shwetaap) [reply travis_ci/test_duplicate_templates.sh]: Added these changes

File: travis_ci/test_duplicate_templates.sh (hunk 6 above)
Comment (shwetaap) [reply travis_ci/test_duplicate_templates.sh:21]: Fixed it

**Comment by omeryahud** [comment]
/lgtm cancel
We figured out that we need to keep the old template, remove os labels from it, and deprecate it

File: templates/windows.tpl.yaml [hunk 13]
```diff
@@ -63,9 +59,6 @@ metadata:
       ]
 
   labels:
-    os.template.kubevirt.io/win2k19: "true"
-    os.template.kubevirt.io/win2k16: "true"
-    os.template.kubevirt.io/win2k12r2: "true"
     workload.template.kubevirt.io/{{ item.workload }}: "true"
```
Comment (omeryahud) [inline templates/windows.tpl.yaml:62]: Remove `workload` and `flavor` labels too please

**Review by omeryahud — CHANGES_REQUESTED** [review]

File: templates/windows.tpl.yaml (hunk 13 above)
Comment (shwetaap) [reply templates/windows.tpl.yaml]: Done

File: templates/windows.tpl.yaml [hunk 14]
```diff
@@ -25,10 +25,6 @@ metadata:
       /objects[0].spec.template.spec.volumes
       /objects[0].spec.template.spec.networks
```
Comment (omeryahud) [inline templates/windows.tpl.yaml:27]: Please add the deprecation annotation like in the win2k12 deprecated template:
https://github.com/kubevirt/common-templates/blob/master/templates/win2k12r2-deprecated.tpl.yaml#L19

File: generate-templates.yaml [hunk 15]
```diff
@@ -232,6 +232,36 @@
     vars:
       osinfoname: win2k12r2
 
+  - name: Generate Windows server 2012 R2 templates
```
Comment (omeryahud) [inline generate-templates.yaml:235]: Change `Generate Windows server templates` to `Generate deprecated Windows server templates` in line 225

File: templates/windows.tpl.yaml (hunk 14 above)
Comment (omeryahud) [reply templates/windows.tpl.yaml:27]: and also please change the name of this file to `deprecated-windows.tpl.yaml` (make sure to change it in `generate-templates.yaml` as well)

**Review by omeryahud — CHANGES_REQUESTED** [review]

File: generate-templates.yaml (hunk 15 above)
Comment (shwetaap) [reply generate-templates.yaml:235]: Done

File: templates/windows.tpl.yaml (hunk 14 above)
Comment (shwetaap) [reply templates/windows.tpl.yaml:27]: Done

File: templates/deprecated-windows.tpl.yaml [hunk 16]
```diff
@@ -25,10 +26,6 @@ metadata:
       /objects[0].spec.template.spec.volumes
       /objects[0].spec.template.spec.networks
 
-    name.os.template.kubevirt.io/win2k19: {{ lookup('osinfo', 'win2k19').name }}
```
Comment (omeryahud) [inline templates/deprecated-windows.tpl.yaml:28]: I found out I was wrong, we shouldn't remove the labels, but we should change their values to `null`, otherwise when the cluster updates the templates, it leaves the labels on if they do not exist in the newer version, setting the label values to `null` would remove them.

File: templates/deprecated-windows.tpl.yaml [hunk 17]
```diff
@@ -63,11 +60,6 @@ metadata:
       ]
 
   labels:
-    os.template.kubevirt.io/win2k19: "true"
```
Comment (omeryahud) [inline templates/deprecated-windows.tpl.yaml:66]: Same for these labels

**Review by omeryahud — CHANGES_REQUESTED** [review]

File: templates/deprecated-windows.tpl.yaml (hunk 17 above)
Comment (shwetaap) [reply templates/deprecated-windows.tpl.yaml:66]: Done

File: templates/deprecated-windows.tpl.yaml (hunk 16 above)
Comment (shwetaap) [reply templates/deprecated-windows.tpl.yaml:28]: Done
